src/lib.rs

```diff
@@ -462,51 +472,47 @@ pub struct DataFormatDescriptor<'data> {
     pub data: &'data [u8],
 }
 
-#[derive(Debug, Copy, Clone)]
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
 pub struct BasicDataFormatDescriptorHeader {
     /// None means Unspecified
     pub color_model: Option<ColorModel>, //: 8;
     /// None means Unspecified
     pub color_primaries: Option<ColorPrimaries>, //: 8;
     /// None means Unspecified
     pub transfer_function: Option<TransferFunction>, //: 8;
-    pub flags: DataFormatFlags,           //: 8;
-    pub texel_block_dimensions: [u32; 4], //: 8 x 4;
-    pub bytes_planes: [u32; 8],           //: 8 x 8;
+    pub flags: DataFormatFlags,                 //: 8;
+    pub texel_block_dimensions: [NonZeroU8; 4], //: 8 x 4;
+    pub bytes_planes: [u8; 8],                  //: 8 x 8;
 }
 
 impl BasicDataFormatDescriptorHeader {
     pub const LENGTH: usize = 16;
 
-    pub fn parse(bytes: &[u8]) -> Result<Self, ParseError> {
-        let mut offset = 0;
+    pub fn as_bytes(&self) -> [u8; Self::LENGTH] {
+        let mut bytes = [0u8; Self::LENGTH];
 
-        let v = bytes_to_u32(bytes, &mut offset)?;
-        let model = shift_and_mask_lower(0, 8, v);
-        let primaries = shift_and_mask_lower(8, 8, v);
-        let transfer = shift_and_mask_lower(16, 8, v);
-        let flags = shift_and_mask_lower(24, 8, v);
+        let color_model = self.color_model.map(|c| c.value()).unwrap_or(0);
+        let color_primaries = self.color_primaries.map(|c| c.value()).unwrap_or(0);
+        let transfer_function = self.transfer_function.map(|t| t.value()).unwrap_or(0);
 
-        let v = bytes_to_u32(bytes, &mut offset)?;
-        let texel_block_dimensions = [
-            shift_and_mask_lower(0, 8, v) + 1,
-            shift_and_mask_lower(8, 8, v) + 1,
-            shift_and_mask_lower(16, 8, v) + 1,
-            shift_and_mask_lower(24, 8, v) + 1,
-        ];
+        let texel_block_dimensions = self.texel_block_dimensions.map(|dim| dim.get() - 1);
 
-        let v = bytes_to_u32(bytes, &mut offset)?;
-        let mut bytes_planes = [0u32; 8];
-        bytes_planes[0] = shift_and_mask_lower(0, 8, v);
-        bytes_planes[1] = shift_and_mask_lower(8, 8, v);
-        bytes_planes[2] = shift_and_mask_lower(16, 8, v);
-        bytes_planes[3] = shift_and_mask_lower(24, 8, v);
+        bytes[0] = color_model;
+        bytes[1] = color_primaries;
+        bytes[2] = transfer_function;
+        bytes[3] = self.flags.bits;
+        bytes[4..8].copy_from_slice(&texel_block_dimensions);
+        bytes[8..16].copy_from_slice(&self.bytes_planes);
 
-        let v = bytes_to_u32(bytes, &mut offset)?;
-        bytes_planes[4] = shift_and_mask_lower(0, 8, v);
-        bytes_planes[5] = shift_and_mask_lower(8, 8, v);
-        bytes_planes[6] = shift_and_mask_lower(16, 8, v);
-        bytes_planes[7] = shift_and_mask_lower(24, 8, v);
+        bytes
+    }
+
+    pub fn from_bytes(bytes: &[u8]) -> Result<Self, ParseError> {
```

Comment:
Should this take a reference to an array, similar to the other existing `from_bytes` functions?

Reply:
Updated. This introduced a couple calls to convert from slices to arrays where we call these methods.